### PR TITLE
feat: adds `api.MAC(..)`

### LINKED ISSUE #416: feat: add `api.AddInPlace()` and `api.MulInPlace`
Issue; when dealing with very large linear expressions, `api.Add` and `api.Mul` perform a lot of memory allocation and memory moves, since the result is always a new variable. If we consider the following snippet though;

```go
// generate the roots of unity <1,ω,ω²,..,ωⁿ⁻¹>
	rous := make([]fr.Element, cardinality)
	rous[0].Set(&cardInverse)
	for i := 1; i < int(cardinality); i++ {
		rous[i].Mul(&rous[i-1], &genInv)
	}
	var acc frontend.Variable
	for i := 0; i < int(cardinality); i++ {
		acc = 0
		for j := 0; j < int(cardinality); j++ {
			e := (j * i) % int(cardinality)
			tmp := api.Mul(rous[e], p[j])
			acc = api.Add(acc, tmp)
		}
		api.AssertIsEqual(acc, res[i])
	}
```

This section: 
```
			tmp := api.Mul(rous[e], p[j])
			acc = api.Add(acc, tmp)
``` 
If `cardinality` is large is going to perform millions of memory allocations and movement, and pre-allocating a linear expression with expected capacity would help tremendously.

### PR DESCRIPTION
**Pull request by gbotrel**
```go
	// MAC sets and return a = a + (b*c)
	// ! may mutate a without allocating a new result
	// ! always use MAC(...) result for correctness
	MAC(a, b, c Variable) Variable
```

Fixes #416 . Impact in `std/math/emulated` for `rsh` is significant (~40% less memallocs).

```go
	for i := 0; i < len(bits); i++ {
		Σbi = api.MAC(Σbi, bits[i], c)
		ΣbiRShift = api.MAC(ΣbiRShift, bits[i], cRShift)

		c.Lsh(c, 1)
		cRShift.Lsh(cRShift, 1)
		api.AssertIsBoolean(bits[i])
	}
```

Also, this new api would result in less constraint in a PlonKish arithmetization, since it will create one constraint instead of 2.

### CODE REVIEW

File: frontend/cs/r1cs/api_assertions.go [hunk 1]
```diff
@@ -83,14 +83,12 @@ func (builder *builder) AssertIsBoolean(i1 frontend.Variable) {
 func (builder *builder) AssertIsLessOrEqual(_v frontend.Variable, bound frontend.Variable) {
 	v := builder.toVariable(_v)
 
-	switch b := bound.(type) {
-	case expr.LinearExpression:
+	if b, ok := bound.(expr.LinearExpression); ok {
```
Comment (ivokub) [inline frontend/cs/r1cs/api_assertions.go:86]: Is it only stylistic change?

File: frontend/cs/r1cs/builder.go [hunk 2]
```diff
@@ -68,6 +72,8 @@ func newBuilder(field *big.Int, config frontend.CompileConfig) *builder {
 		mtBooleans: make(map[uint64][]expr.LinearExpression, config.Capacity/10),
 		config:     config,
 		heap:       make(minHeap, 0, 100),
+		mbuf1:      make(expr.LinearExpression, 0, 100),
```
Comment (ivokub) [inline frontend/cs/r1cs/builder.go:75]: If CompressThreshold is set in config then it would be better to use this as the initial capacity of the buffers?

**Review by ivokub — CHANGES_REQUESTED** [review]
Looks good. I think it is a clean and required change.

File: frontend/cs/r1cs/api.go [hunk 3]
```diff
@@ -40,19 +40,69 @@ import (
...
 
+		if v1Constant {
+			builder.mbuf1 = append(builder.mbuf1, builder.toVariable(c)...)
+			builder.mulConstant(builder.mbuf1, n1, true)
+			return
+		}
+		builder.mbuf1 = append(builder.mbuf1, builder.toVariable(b)...)
+		builder.mulConstant(builder.mbuf1, n2, true)
+	}
+	mulBC()
```
[the remark this answers is not in the document]
Comment (gbotrel) [reply frontend/cs/r1cs/api.go:78]: yes, also almost a copy/paste from the `api.Mul` didn't want to rewrite struct and introduce bugs.

File: frontend/cs/r1cs/api_assertions.go (hunk 1 above)
Comment (gbotrel) [reply frontend/cs/r1cs/api_assertions.go:86]: yes -- dirty history, my bad. (one attempt I add `api.MAC` take a pointer as a parameter to grow the linear expression, but it complicated things in too many places)

File: std/math/emulated/wrapped_api.go [hunk 4]
```diff
@@ -76,6 +76,11 @@ func (w *FieldAPI[T]) Add(i1 frontend.Variable, i2 frontend.Variable, in ...fron
 	return res
 }
 
+func (w *FieldAPI[T]) MAC(a, b, c frontend.Variable) frontend.Variable {
+	// TODO can we do better here to limit allocations?
+	return w.Add(a, w.Mul(b, c))
```
[the remark this answers is not in the document]
Comment (gbotrel) [reply std/math/emulated/wrapped_api.go:81]: I'll let you tackle that one, feel free to push to this branch -- each time I touch the `emulated` package I break things.

File: frontend/cs/scs/api.go [hunk 5]
```diff
@@ -52,6 +52,12 @@ func (builder *scs) Add(i1, i2 frontend.Variable, in ...frontend.Variable) front
 
 }
 
+func (builder *scs) MAC(a, b, c frontend.Variable) frontend.Variable {
+	// TODO can we do better here to limit allocations?
+	// technically we could do that in one PlonK constraint (against 2 for separate Add & Mul)
+	return builder.Add(a, builder.Mul(b, c))
```
[the remark this answers is not in the document]
Comment (gbotrel) [reply frontend/cs/scs/api.go:58]: Am going to wait on @ThomasPiellard new backend on that one; plonk frontend is in a need of a refactor to convert input to term, and add custom constraints etc... keeping the TODO around for now.

**Comment by ivokub** [comment]
Suggested edit:
```diff
diff --git a/std/math/emulated/wrapped_api.go b/std/math/emulated/wrapped_api.go
index aa433d24..6d40f4ee 100644
--- a/std/math/emulated/wrapped_api.go
+++ b/std/math/emulated/wrapped_api.go
@@ -77,8 +77,10 @@ func (w *FieldAPI[T]) Add(i1 frontend.Variable, i2 frontend.Variable, in ...fron
 }
 
 func (w *FieldAPI[T]) MulAcc(a, b, c frontend.Variable) frontend.Variable {
-	// TODO can we do better here to limit allocations?
-	return w.Add(a, w.Mul(b, c))
+	els := w.varsToElements(a, b, c)
+	mul := w.f.Mul(els[1], els[2])
+	res := w.f.Add(els[0], mul)
+	return res
 }
 
 func (w *FieldAPI[T]) Neg(i1 frontend.Variable) frontend.Variable {

```

**Comment by ivokub** [comment]
> Suggested edit:
> 
> ```diff
> diff --git a/std/math/emulated/wrapped_api.go b/std/math/emulated/wrapped_api.go
> index aa433d24..6d40f4ee 100644
> --- a/std/math/emulated/wrapped_api.go
> +++ b/std/math/emulated/wrapped_api.go
> @@ -77,8 +77,10 @@ func (w *FieldAPI[T]) Add(i1 frontend.Variable, i2 frontend.Variable, in ...fron
>  }
>  
>  func (w *FieldAPI[T]) MulAcc(a, b, c frontend.Variable) frontend.Variable {
> -	// TODO can we do better here to limit allocations?
> -	return w.Add(a, w.Mul(b, c))
> +	els := w.varsToElements(a, b, c)
> +	mul := w.f.Mul(els[1], els[2])
> +	res := w.f.Add(els[0], mul)
> +	return res
>  }
>  
>  func (w *FieldAPI[T]) Neg(i1 frontend.Variable) frontend.Variable {
> ```

Tried to see if could reduce allocating new `Element`, but seems to be larger overhaul to have methods on `Field` type which also take the destination. I think it makes sense to do it separate PR.

**Pull request merged**